linkerd/app/gateway/src/lib.rs

```diff
@@ -85,161 +93,246 @@ where
         dispatch_timeout,
         ..
     } = inbound.config().proxy.clone();
+    let local_id = inbound.runtime().identity.as_ref().map(|l| l.id().clone());
+
+    // For each gatewayed connection that is *not* HTTP, use the target from the
+    // transport header to lookup a service profile. If the profile includes a
+    // resolvable service name, then continue with TCP endpoint resolution,
+    // balancing, and forwarding. An invalid original destination address is
+    // used so that service discovery is *required* to provide a valid endpoint.
+    //
+    // TODO: We should use another target type that actually reflects
+    // reality. But the outbound stack is currently pretty tightly
+    // coupled to its target types.
+    let tcp = outbound
+        .clone()
+        .push_tcp_endpoint()
+        .push_tcp_logical(resolve.clone())
+        .into_stack()
+        .push_request_filter(|(p, _): (Option<profiles::Receiver>, _)| match p {
+            Some(rx) if rx.borrow().name.is_some() => Ok(outbound::tcp::Logical {
+                profile: Some(rx),
+                orig_dst: std::net::SocketAddr::from(([0, 0, 0, 0], 0)),
+                protocol: (),
+            }),
+            _ => Err(discovery_rejected()),
+        })
+        .push(profiles::discover::layer(profiles.clone(), {
+            let allow = allow_discovery.clone();
+            move |addr: NameAddr| {
+                if allow.matches(addr.name()) {
+                    Ok(addr)
+                } else {
+                    Err(RefusedNotResolved(addr))
+                }
+            }
+        }))
+        .push_on_response(
+            svc::layers()
+                .push(
+                    inbound
+                        .runtime()
+                        .metrics
+                        .stack
+                        .layer(metrics::StackLabels::inbound("tcp", "gateway")),
```

Comment:
nit/TIOLI: i wonder if we're writing stuff like this in enough places that it would be good for the `Inbound` (and `Outbound`) types to have a method, so we could just say
```rust
    .push(inbound.stack_metrics("tcp", "gateway"))
```

not a big deal though.